app/domain/authentication/authn_k8s/kubectl_exec.rb

```diff
@@ -1,133 +1,214 @@
+            logger.debug("Pod: #{pod_name}, message: close, data: #{messages.msg_data(msg)}")
             close
           end
         end
 
-        ws.on :message do |msg|
-          __send__ "message_#{msg.type}", msg
-        end
-
         ws.on :open do
           hs = ws.handshake
+
           if hs.error
-            emit :error, hs.instance_variable_get("@data")
+            emit(:error, messages.messages)
           else
-            Rails.logger.debug "Pod #{@pod.metadata.name.inspect} : channel open"
+            logger.debug("Pod #{pod_name} : channel open")
 
             if stdin
-              ws.send(STDIN_CHANNEL.chr + body)
-              ws.send nil, :type => :close
+              data = messages.channel('stdin').chr + body
+              ws.send_msg(data)
+              ws.send_msg(nil, type: :close)
             end
           end
         end
 
-        ws.on :close do |e|
-          @closed = true
-          Rails.logger.debug "Pod #{@pod.metadata.name.inspect} : channel closed"
+        ws.on(:close) do |e|
+          stream_state.close
+          logger.debug("Pod #{pod_name} : channel closed")
         end
 
-        ws.on :error do |e|
-          @closed = true
-          Rails.logger.debug "Pod #{@pod.metadata.name.inspect} error : #{e.inspect}"
-          @messages[:error] << e.inspect
+        ws.on(:error) do |e|
+          stream_state.close
+          logger.debug("Pod #{pod_name} error : #{e.inspect}")
+          
+          messages.save_string(e.inspect, stream: :error)
         end
+      end
 
-        ws.send "\n"
-
+      def wait_for_close_message(ws)
         (@timeout / 0.1).to_i.times do
-          break if ws.instance_variable_get("@closed")
+          break if @stream_state.closed?
           sleep 0.1
         end
+      end
 
-        if ws.instance_variable_get("@closed")
-          ws.instance_variable_get "@messages"
-        else
-          raise "Command timed out in container #{container.inspect} of Pod #{@pod.metadata.name.inspect}"
-        end
+      def query_string(cmds, stdin)
+        stdin_part = stdin ? ['stdin=true'] : []
+        cmds_part = cmds.map { |c| "command=#{CGI.escape(c)}" }
+        (base_query_string_parts + stdin_part + cmds_part).join("&")
       end
 
-      def copy path, content, mode
-        exec [ 'tar', 'xvf', '-', '-C', '/' ], stdin: true, body: generate_file_tar_string(path, content, mode)
+      def base_query_string_parts
+        [ "container=#{CGI.escape(@container)}", "stderr=true", "stdout=true" ]
       end
 
-      private
+      def server_url(cmds, stdin)
+        api_uri = @kubeclient.api_endpoint
+        base_url = "wss://#{api_uri.host}:#{api_uri.port}"
+        path = "/api/v1/namespaces/#{@pod_namespace}/pods/#{@pod_name}/exec"
+        query = query_string(cmds, stdin)
+        "#{base_url}#{path}?#{query}"
+      end
 
-      def generate_file_tar_string path, content, mode
+      def generate_file_tar_string(path, content, mode)
```

Comment:
We should fix this name, it's a carry over from the old code and doesn't really make sense...

Reply:
afaik this is still doing exactly what it says. What would you call it?

Reply:
Changing to `tar_file_as_string` from Slack convo.